### ISSUE TITLE
How to update or reload env variables ?

### ISSUE THREAD

**Issue by tlartaud** [motdotla/dotenv#122]
Hi @motdotla

I'm using Grunt and dotenv.
At the beginning of my Gruntfile, I have
```
    var dotenv = require('dotenv');

    // Load DotEnv variables
    dotenv.load();
```

I have a task using grunt-text-replace that replace some variable inside my .env file from users prompts. After the file .env has been edited by my grunt task, i want the variables to be reloaded again, but if call `dotenv.load();` again, it seems it has no effect and the old variables are used by my application.

Is there another way to update the .env file, and being able to use the new variables with `process.env.MYVAR`

Basically, here is an example of what i want to do : Imagine the following task
- clone project repository
- grunt update_db_credentials
  - load default .env file from project repository
  - load task that prompt the user for new db credentials
  - store the db credentials in the .env file
  - reload the .env file so the var can be accessed directly using `process.env.DB_NAME`

Thanks a lot for your help.

**Comment by maxbeatty** [comment]
You'll need to restart the _process_ or spawn a new one.

I wrote [a similar setup script](https://github.com/jsperf/jsperf.com/blob/master/setup/index.js) and use the values from the prompt instead of `process.env` to continue setting up.

**Comment by tlartaud** [comment]
@maxbeatty i've tried to create a grunt task where i call `dotenv.load();` for the second time, after the .env file has been edited. `process.env.DB_NAME` still return the variable from the default .env file :(

**Comment by maxbeatty** [comment]
Any existing value in `process.env` is respected over any value from a .env file ([src](https://github.com/motdotla/dotenv/blob/master/lib/main.js#L33)). Calling `load` or `config` twice in the same script will not result in any changes because `process.env` is already populated. I would suggest using the inputs a user gives you directly or creating a separate task to use the new values instead of one task to do both.

**Comment by tlartaud** [comment]
OK thanks for this information !

**Comment by jlpb97** [comment]
**f_ff_****f****f**

_f_

**Comment by apoorva-shah** [comment]
> You'll need to restart the _process_ or spawn a new one.
> 
> I wrote [a similar setup script](https://github.com/jsperf/jsperf.com/blob/master/setup/index.js) and use the values from the prompt instead of `process.env` to continue setting up.

Hello this link is not working. Can you please paste it again?

My .env file has
PROJECT_ENV=development

My file is 1.js. Code is as below

const dotenv = require('dotenv');
dotenv.config();
console.log(process.env.PROJECT_ENV); // I always get development here. i want test when i run 2nd time.
process.env.PROJECT_ENV='test'; // I am changing env variable here
console.log(process.env.PROJECT_ENV);

So every time(from 2nd time run) i run this file i should get 'test' as console.log i always get development in result.

Thanks in advance.

**Comment by valmack** [comment]
Use something like inquirer or parse process.argv

```
/**
 * Returns user input.
 * @param {string} message Message for question
 * Throws if no valid input is found.
 */
export async function getUserInput(message: string) {
  const questions = [
    {
      type: "input",
      name: "input",
      message: message,
    },
  ];
  const inputs = await inquirer.prompt(questions);
  if (!inputs.input || inputs.input == "") {
    throw new Error("Invalid input");
  } else {
    return inputs.input;
  }
}
```

**Comment by Thejesh-404** [comment]
> > You'll need to restart the _process_ or spawn a new one.
> > I wrote [a similar setup script](https://github.com/jsperf/jsperf.com/blob/master/setup/index.js) and use the values from the prompt instead of `process.env` to continue setting up.
> 
> Hello this link is not working. Can you please paste it again?
> 
> My .env file has
> PROJECT_ENV=development
> 
> My file is 1.js. Code is as below
> 
> const dotenv = require('dotenv');
> dotenv.config();
> console.log(process.env.PROJECT_ENV); // I always get development here. i want test when i run 2nd time.
> process.env.PROJECT_ENV='test'; // I am changing env variable here
> console.log(process.env.PROJECT_ENV);
> 
> So every time(from 2nd time run) i run this file i should get 'test' as console.log i always get development in result.
> 
> Thanks in advance.


I'm working on the similar script
got any solutions?

**Comment by WeslyG** [comment]
I struggled for a long time over the problem of restarting my nestjs application (node js) when changing the dotenv 
configuration, and found a cool solution.

```
new NodemonPlugin({
      watch: path.resolve('.env'),
      nodeArgs: ['--require=dotenv/config']
    }),
```

I disabled the webpack-dotenv plugin because it inline my variables into the final build.

using nodemon, I monitor the dotenv file, and when it changes, I restart the build by executing the config function from the dotenv package directly in the nodemon plugin.
If you use webpack and nodejs (backend) then it works very well

**Comment by modscleo4** [comment]
Run `delete process.env.VARIABLE` before `dotenv.config()`. It will reload the VARIABLE from .env

**Comment by motdotla** [comment]
You can now use the `override` option to solve for this:

https://github.com/motdotla/dotenv#override

**Comment by przemek-nowicki** [comment]
If someone would like to update env variables but does not use dotenv library to access env variable, it's possible to signal the application main process by other process. Here is the simple lib I wrote to archive it with an example how to do it:

**Comment by RyadPasha** [comment]
Here's a straightforward and clean function that allows you to reload all the .env variables after pulling them without the need to restart the node process:

```
/**
 * @author Mohamed Riyad <m@ryad.me>
 *
 * This function will reload the .env variables.
 */
const reloadEnv = () => {
    const envConfig = dotenv.parse(fs.readFileSync('.env'))

    for (const key in envConfig) {
        process.env[key] = envConfig[key]
    }
}

**Comment by motdotla** [comment]
nice clean solution @RyadPasha

**Comment by codexbits** [comment]
> Here's a straightforward and clean function that allows you to reload all the .env variables after pulling them without the need to restart the node process:
> 
> ```
> const fs = require('fs');
> const dotenv = require('dotenv');
> 
> // Load initial .env variables
> dotenv.config();
> 
> /**
>  * @author Mohamed Riyad <m@ryad.me>
>  *
>  * This function will reload the .env variables.
>  */
> const reloadEnv = () => {
>     const envConfig = dotenv.parse(fs.readFileSync('.env'))
> 
>     for (const key in envConfig) {
>         process.env[key] = envConfig[key]
>     }
> }
> 
> // Reload .env variables
> reloadEnv();
> ```

This fixed my application not updating the correct hostname! Thank you!

**Comment by ocampoje17** [comment]
> You can now use the `override` option to solve for this:
> 
> https://github.com/motdotla/dotenv#override

Confirm this work well

**Comment by motdotla** [comment]
for anyone who comes across this in the future you can also use [dotenvx](https://github.com/dotenvx/dotenvx). It has override/overload and many more features.
